sdk/cosmos/azure-cosmos-kafka-connect/src/main/java/com/azure/cosmos/kafka/connect/implementation/source/MetadataCosmosStorageManager.java

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.kafka.connect.implementation.source;
+
+import com.azure.cosmos.CosmosAsyncContainer;
+import com.azure.cosmos.implementation.Utils;
+import com.azure.cosmos.implementation.apachecommons.lang.tuple.Pair;
+import com.azure.cosmos.kafka.connect.implementation.KafkaCosmosExceptionsHelper;
+import com.azure.cosmos.models.PartitionKey;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import reactor.core.publisher.Mono;
+
+import java.util.Map;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class MetadataCosmosStorageManager implements IMetadataReader {
+    private final CosmosAsyncContainer metadataContainer;
+
+    public MetadataCosmosStorageManager(CosmosAsyncContainer metadataContainer) {
+        checkNotNull(metadataContainer, "Argument 'metadataContainer' can not be null");
+        this.metadataContainer = metadataContainer;
+    }
+
+    public void createMetadataItems(MetadataTaskUnit metadataTaskUnit) {
+        // To be as consistent as what will be persisted to kafka topic if kafka being the storage type
+        // we created the cosmos metadata item from the kafka topic records
+
+        // add the containers metadata items - it tracks the databaseName -> List[containerRid] mapping
+        this.createContainersMetadataItem(
+            metadataTaskUnit.getContainersMetadata().getLeft(),
+            metadataTaskUnit.getContainersMetadata().getRight());
+
+        // add the container feedRanges metadata item - it tracks the containerRid -> List[FeedRange] mapping
+        for (Pair<FeedRangesMetadataTopicPartition, FeedRangesMetadataTopicOffset> feedRangesMetadata : metadataTaskUnit.getFeedRangesMetadataList()) {
+            this.createFeedRangesMetadataItem(feedRangesMetadata.getLeft(), feedRangesMetadata.getRight());
+        }
+    }
+
+    public void createContainersMetadataItem(
+        ContainersMetadataTopicPartition topicPartition,
+        ContainersMetadataTopicOffset topicOffset) {
+
+        ContainersMetadataItem containersMetadataItem =
+            new ContainersMetadataItem(
+                topicPartition.getDatabaseName(),
+                ContainersMetadataTopicOffset.toMap(topicOffset));
+
+        this.metadataContainer.upsertItem(containersMetadataItem).block();
```

Comment:
converting into returnableException for transient exception, apply to other methods as well in this PR